docs/syntax.md

```diff
@@ -0,0 +1,667 @@
+The exact function they become depends on the encoding.
+
+## Statements
+
+### Assignment
+
+```python
+value = 2
+return value
+
+(first, second) = (1, 2)
+return second
+
+{x y} = {2 3}
+```
+
+Assigns a value to a variable, it's possible to pattern match match tuples and superpositions.
+
+### In-Place Operation
+
+```python
+x += 1
+return x
+```
+
+The in-place operation does an infix operation and re-assigns a variable.
+
+The operations are:
+- Addition `+=`
+- Subtraction `-=`
+- Multiplication `*=`
+- Division `/=`
+
+### Return
+
+```python
+return "hello"
+```
+
+Returns the following expression. All paths or branches should end with a return.
+
+### If
+
+```python
+if condition:
+  return 0
+else:
+  return 1
+```
+
+A branching statement where `else` is mandatory.
+
+The condition must return an `u24` and it is equivalent to a switch statement:
+
+```
+switch _ = condition:
+  0: else
+  _: then
+```
+
+### Switch
+
+```python
+switch x = 4:
+  0: "Zero"
+  1: "One"
+  _: "Not zero or one"
+```
+
+A switch for native numbers, the pattern matching cases must start from `0` up to `_` sequentially.
+
+It is possible to bind a variable name to the matching value, it allows the access to the predecessor `x-2` (in this case) or `bound_var-next_num` (in the general case).
+
+### Match
+
+```python
+match x = Option/none:
+  Option/some:
+    return x.value
+  Option/none:
+    return 0
+```
+
+A pattern matching statement, the cases must be the constructor names of the matching value.
+
+It is possible to bind a variable name to the matching value. The fields of the matched constructor are bound to `matched_var.field_name`.
+
+### Fold
+
+```python
+fold x = Tree/leaf:
+  Tree/node:
+    return x.value + x.left + x.right
+  Tree/leaf:
+    return 0
+```
+
+A fold statement. Reduces the given value with the given match cases.
+
+It is possible to bind a variable name to the matching value. Just like in `match`, the fields are bound to `matched_var.field_name`.
+
+For fields notated with `~` in the type definition, the fold function is called implicitly.
+
+It is equivalent to the inline recursive function:
+
+```python
+def fold(x):
+  match x:
+    Tree/Node:
+      return x.value + fold(x.left) + fold(x.right)
+    Tree/Leaf:
+      return 0
+...
+fold(Tree/Leaf)
+```
+
+### Bend
+
+Bend can be used to create recursive data structures:
+
+```rust
+bend x = 0 while x < 10:
+  left = go(x + 1)
+  right = go(x + 1)
+  result = Tree/Node(left, right)
+then:
+  result = Tree/Leaf(x)
+```
+
+Which binds a variable to the return of an inline recursive function.
+The function `go` is available inside the bend body and calls it recursively.
+
+It is possible to initialize multiple variables:
+
+```python
+bend x = 1, y = 2 ... while condition(x, y, ...):
+```
+
+It is equivalent to this inline recursive function:
+
+```python
+def bend(x, y, ...):
+  if condition(x, y, ...):
+    ...
+    return ... bend(x, y, ...) ...
```

Comment:
Which keyword are we using to indicate the recursive call? `bend` or `go`? I think both are ok. `bend` is more cool looking, but `go` might be more intuitive.
`rec` would be good for people who are used to languages that have a similar marker to indicate recursive functions, but I suspect that's not a lot of people.